olive/passes/onnx/perf_tuning.py

```diff
@@ -163,99 +169,153 @@ def threads_num_tuning(model, data_root, latency_metric, config, tuning_combo):
     elif test_params["session_options"].get("execution_mode") == ort.ExecutionMode.ORT_SEQUENTIAL:
         threads_names = ["intra_op_num_threads"]
     else:
         threads_names = ["inter_op_num_threads", "intra_op_num_threads"]
-    best_latency = None
 
-    if test_params["session_options"].get("inter_op_num_threads") and test_params["session_options"].get(
-        "intra_op_num_threads"
+    tuning_results = []
+
+    if (
+        test_params["session_options"].get("inter_op_num_threads") is not None
+        and test_params["session_options"].get("intra_op_num_threads") is not None
     ):
-        test_result = get_benchmark(model, data_root, latency_metric, config, test_params)
+        # If user specify both inter_op_num_threads and intra_op_num_threads, we will not do tuning
+        test_result = get_benchmark(model, data_root, latency_metric, config, test_params, io_bind)
         tuning_results.append(test_result)
-        best_latency = test_result["latency_ms"]
-    else:
-        for threads_name in threads_names:
-            thread_num = test_params["session_options"].get(threads_name)
-            if thread_num:
-                upper_threads_num = thread_num
-                lower_threads_num = thread_num
+        return tuning_results
+
+    for threads_name in threads_names:
+        # set the upper bound and lower bound for binary search
+        thread_num = test_params["session_options"].get(threads_name)
+        if thread_num is not None:
+            upper_threads_num = thread_num
+            lower_threads_num = thread_num
+        else:
+            upper_threads_num = config.cpu_cores or psutil.cpu_count(logical=False)
+            lower_threads_num = 1
+
+        current_threads_num = lower_threads_num
+        best_latency = None
+        best_threads_num = None
+
+        while lower_threads_num < upper_threads_num:
+            test_params["session_options"][threads_name] = current_threads_num
+            test_result = get_benchmark(model, data_root, latency_metric, config, test_params, io_bind)
+            tuning_results.append(test_result)
+
+            if best_latency is None:
+                # the first time run benchmark, then change next to the upper bound
+                best_latency = test_result["latency_ms"]
+                best_threads_num = current_threads_num
+                current_threads_num = upper_threads_num
+            elif best_latency < test_result["latency_ms"]:
+                mid_threads_num = lower_threads_num + (upper_threads_num - lower_threads_num) // 2
+                # the current benchmark is worse than last time.
+                # Just keep the best_latency and best_threads_num
+                if best_threads_num < current_threads_num:
+                    # update the upper bound to middle if last time is in lower side.
+                    upper_threads_num = mid_threads_num
+                    next_thread_num = upper_threads_num
```

Comment:
Maybe no need to intro a new variable? Directly use `current_thread_num` is fine, just to ensure update best_latency before change the `current_thread_num`.

Reply:
I add the variable to make the assignment be consistent with the line 256.